### DISCUSSION TITLE
2/26/21 iOS build lunch notes

### DISCUSSION THREAD

**Discussion by keith** [MobileNativeFoundation/discussions#7] [General]
Context: Every once and a while there is an iOS tooling meetup, these are the notes from the most recent one.

* [Jerry] [Index while building V2 architecture](https://github.com/bazel-ios/rules_ios/blob/master/docs/index_while_building.md) and key discoveries - 10 min lightning talk
   * https://github.com/bazel-ios/rules_ios/blob/master/docs/index_while_building.md
   * Share a global index store (like Xcode does)
   * Overhead is now down to 3-5% (instead of 300% for separate index stores per library)
   * An aspect can run while the build is running (or BEP program) to show index data to users while it’s being calculated
   * How much extra work needs to happen going up & back to/from the remote cache? Just paths?
      * In the current implementation clang and Swift write absolute paths into the unit files. So you could work around by either matching directories across CI and local development, or by writing absolute paths to CI cache and then local jobs can re-write the absolute paths it fetches from the cache.
   * Does lock contention become an issue?
      * No so far.
   * With remote execution, would this global index store cache need to be maintained across all machines involved in the build?
      * https://github.com/bazelbuild/rules_swift/issues/561
   * Are there complications with Objective-C for this approach?
      * The Swift rules are already outputting an index store directory for these rules
      * This isn’t implemented in rules-ios yet, thus the complications. Could be some sort of aspect that would help with this.
* [Matthew] Next generation iOS build & test at Robinhood - 10m
   * Had been using hosted machines from Buildkite, was really fast
   * Looking at running iOS ecosystem on AWS Graviton (generic ARM)
   * https://github.com/corellium/corellium-api
   * https://github.com/Flank/flank/pull/1621
   * https://robinhood.engineering/increasing-efficiency-confidence-with-project-apollo-cea1ebe0f282
   * Madhuri (https://www.elotl.co/) will share benchmark numbers later
* [Madhuri] Community interest in learning about build benchmarking numbers on AWS mac1.metal
   * Benchmark experiment: build time for building Robinhood iOS app
      * Status quo: Mac Mini + Parallels + Buildkite. Versus,
      * AWS: mac1.metal + buildkite
         * without virtualization
         * with Parallels
         * with Anka
         * with kubernetes kubelet process wrapper
   * Running these tests for Robinhood
* [Brian / Shawn] quick updates on Bazel Remote Execution
   * Using build farm for remote execution. Works in their datacenter. 19 Mac Mini machines, 1 for compilation.
   * MacBook Pro kills the GRPC connection whenever a build is attempted to start.
   * Investigating what kills GRPC. Anyone else seen this?
   * Bazel throws a Java IO exception.
   * Works fine from Mac mini machines in the datacenter (could be VPN or endpoint detection software)
   * Currently slower than a single machine locally, but many more knobs to turn to tune and benchmark
   * Explicit modules for Swift will be a requirement to make this work well
      * Need a certain rule that will set up the dependency tree for these modules (Lyft can potentially upstream later)
   * You can now use both the disk cache and the remote cache (wasn’t the case before). Also, by proxying you suffer the serialization/deserialization cost extra times.
   * Local disk cache doesn’t (currently) support size limits
   * Enable compression after the edge nodes (cuts network traffic and disk usage by 60-70%). Gradle does this.
* [Jeff] How deeply to teams use implementation only (imports)
   * Private to your module
   * Prevent deep header import trees

**Comment by BalestraPatrick** [comment]
@[Brian / Shawn] We saw a similar problem where our VPN client was blocking gRPC calls through Ipv6, so we had to modify the config to allow that. Maybe your problem is similar? https://github.com/bazelbuild/bazel/issues/3602#issuecomment-633219041

(side note: would be nice to collect people's GitHub usernames in order to mention them directly here)

**Reply by bdittmer** [reply]
Thanks for the tip @BalestraPatrick! We'll look into that and report back.

**Reply by bdittmer** [reply]
Hey @BalestraPatrick -- similar issue on our end. The firewall identifies the traffic as malicious and shuts it down. Remote Execution works just fine if we do it over an SSH tunnel!

**Comment by donnie-jp** [comment]
Hey 👋  
How can I find out about/join the lunch meeting? Interesting topics!

**Reply by freak4pc** [reply]
Interested as well, if it's in any way open to the public :)

**Reply by jshchnz** [reply]
also interested!

**Reply by tinder-maxwellelliott** [reply]
I am also interested!

**Reply by Vkt0r** [reply]
Same here, interesting topics.
